contracts/NFTStorefront.cdc

```diff
@@ -233,22 +266,33 @@ pub contract NFTStorefront {
         pub fun getDetails(): ListingDetails {
             return self.details
         }
+
         
         // purchase
         // Purchase the listing, buying the token.
-        // This pays the beneficiaries and returns the token to the buyer.
-        //
-        pub fun purchase(payment: @FungibleToken.Vault): @NonFungibleToken.NFT {
+        // This pays the beneficiaries and commission to the facilitator and returns extra token to the buyer.
+        // This also cleans up existing listings for the item being purchased.
+        pub fun purchase(
+            payment: @FungibleToken.Vault, 
+            commissionRecipient: &AnyResource{FungibleToken.Receiver},
+        ): @NonFungibleToken.NFT {
+
             pre {
                 self.details.purchased == false: "listing has already been purchased"
                 payment.isInstance(self.details.salePaymentVaultType): "payment vault is not requested fungible token"
                 payment.balance == self.details.salePrice: "payment vault does not contain requested price"
+                self.details.expiry > getCurrentBlock().timestamp: "Listing is expired"
+                self.owner != nil : "Resource doesn't have the assigned owner"
```

Comment:
```suggestion
                self.owner == nil : "Resource doesn't have the assigned owner"
```

if the resource doesn't have the owner... wouldn't the owner be nil?

Reply:
But we want that resource should have the owner so we check that it should not be `nil` otherwise receive an error.